### ISSUE TITLE
Failed to inference type for await inside a impl block with GATs

### ISSUE THREAD

**Issue by xxchan** [rust-lang/rust-analyzer#13705]
**rust-analyzer version**: 0.4.1304-standalone (398a71aff 2022-11-29)

**rustc version**: rustc 1.66.0-nightly (b8c35ca26 2022-10-15)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

```rust
struct T<S> {
    s: S,
}

trait TT: GAT1 {
    type Local: GAT1;
}

pub trait GAT1 {
    type GAT<'a>: Future<Output = ()> + Send + 'a;
}

pub trait GAT2 {
    type GAT<'a>: Future<Output = ()> + Send + 'a;
}

impl<S: TT> T<S> {
    async fn foo() {
        let a = bar();    
        let a = bar().await;

    }
}

async fn bar() -> usize {
    1
}

```

The second one's type cannot be inferred.

<img width="416" alt="image" src="https://user-images.githubusercontent.com/37948597/205286878-a14a77f0-bf4f-40c8-a8b8-8d99940dd00f.png">

Note: If we change one of the GAT1 to GAT2, it works.
```rust
trait TT: GAT1 {
    type Local: GAT1;
}
```

**Comment by xxchan** [comment]
Another observation: If we change the output of `Super1::I` to the same type as `bar`, it works.
```rust
pub trait Super1 {
    type I: std::future::Future<Output = usize> ;
}
```

If we change it to another one, e.g., `u32`, it doesn't work again.

**Comment by xxchan** [comment]
More interestingly, `await` on any identifier that is not an `async fn` will get `Super1::I::Output`

<img width="723" alt="image" src="https://user-images.githubusercontent.com/37948597/205490958-fef8e0cf-8aa2-42a3-98ad-42995db9e564.png">


<img width="488" alt="image" src="https://user-images.githubusercontent.com/37948597/205490994-019e4a56-4180-4553-a5e7-a219f4afea72.png">

**Comment by flodiebold** [comment]
Yeah, pretty sure this is a chalk bug, we have a few similar ones, e.g. #10312; it might be a variation of https://github.com/rust-lang/chalk/issues/727.

**Comment by ChayimFriedman2** [comment]
Working now.
